redis/server.cc

```diff
@@ -0,0 +1,259 @@
+{
+    ++_server._total_connections;
+    ++_server._current_connections;
+    _server._connections_list.push_back(*this);
+}
+
+redis_server::connection::~connection() {
+    --_server._current_connections;
+    _server._connections_list.erase(_server._connections_list.iterator_to(*this));
+    _server.maybe_idle();
+}
+
+future<> redis_server::connection::process()
+{
+    return do_until([this] {
+        return _read_buf.eof();
+    }, [this] {
+        return with_gate(_pending_requests_gate, [this] {
+            return process_request();
+        });
+    }).then_wrapped([this] (future<> f) {
+        try {
+            f.get();
+        }
+        catch (redis_exception& e) {
+        logging.info(" redis exception = {} ", e.what());
+            write_reply(e);
+        }
+        catch (std::exception& e) {
+        logging.info(" std::exception = {} ", e.what());
+            write_reply(redis_exception { e.what() });
+        }
+        catch (...) {
+        logging.info(" ... exception ");
+            write_reply(redis_exception { "Unknown exception" });
+        }
+    }).finally([this] {
+        return _pending_requests_gate.close().then([this] {
+            return _ready_to_respond.finally([this] {
+                return _write_buf.close();
+            });
+        });
+    });
+}
+
+future<> redis_server::connection::shutdown()
+{
+    try {
+        _fd.shutdown_input();
+        _fd.shutdown_output();
+    } catch (...) {
+    }
+    return make_ready_future<>();
+}
+
+thread_local redis_server::connection::execution_stage_type redis_server::connection::_process_request_stage {"redis_transport", &connection::process_request_one};
+
+future<redis_server::result> redis_server::connection::process_request_internal() {
+    return _process_request_stage(this, std::move(_parser.get_request()), seastar::ref(_options), empty_service_permit());
+}
+
+void redis_server::connection::write_reply(const redis_exception& e)
+{
+    _ready_to_respond = _ready_to_respond.then([this, exception_message = e.what_message()] () mutable {
+        return redis_message::make_exception(exception_message).then([this] (auto&& result) {
+            auto message = result.message();
+            return _write_buf.write(std::move(*message)).then([this] {
+                return _write_buf.flush();
+            });
+        });
+    });
+}
+
+future<> redis_server::connection::process_request() {
+    _parser.init();
+    return _read_buf.consume(_parser).then([this] {
+        ++_server._requests_served;
+        ++_server._requests_serving;
+        _pending_requests_gate.enter();
+        auto leave = defer([this] { _pending_requests_gate.leave(); });
+        return process_request_internal().then([this, leave = std::move(leave)] (auto&& result) mutable {
+            --_server._requests_serving;
+            auto message = result.make_message();
+            return _write_buf.write(std::move(*message)).then([this] {
```

Comment:
I'm not sure I understood the purpose of the pending_request_gate - don't you need to also put "leave" into this continuation so the gate keeps getting taken, so we don't kill the request before we finished writing its response? Or is that fine?